src/test/java/api/requests/RequestByInstanceIdResourceTests.java

```diff
@@ -0,0 +1,230 @@
+package api.requests;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.net.MalformedURLException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.RequestType;
+import org.folio.circulation.domain.representations.RequestByInstanceIdRequest;
+import org.folio.circulation.resources.RequestByInstanceIdResource;
+import org.folio.circulation.support.Pair;
+import org.folio.circulation.support.Result;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.ISODateTimeFormat;
+import org.junit.Test;
+
+import api.support.APITests;
+import api.support.builders.LocationBuilder;
+import api.support.fixtures.ItemExamples;
+import io.vertx.core.json.JsonObject;
+
+public class RequestByInstanceIdResourceTests extends APITests {
+
+  @Test
+  public void canTransformInstanceToItemRequests(){
```

Comment:
I'm finding this test difficult to follow. Can you help me understand this test better.

What does this test do? It seems to create multiple requests, which is confusing to me,  thought an instance level request maps to a single item level request?

What do `i` and `j` represent? What does the loop do?

I find that branching (e.g. `if`) statements in tests tends to reduce readability and in some cases can suggest non-deterministic behaviour, which leads to brittle tests.

Reply:
This test is basically a unit test that checks the step where it creates _n_*_instance_ total number of item requests. Recall that  it does not know which type of request to make, so it creates all possible requests for all request types for that one item. This test loops through two items using the **j** counter and for each item it loops through all possible request types with the **i** counter.

Other tests in this file needed to make use of various fixtures so this test class has to extend APITests. We can create another class for true unit-level tests if needed.

Reply:
> This test is basically a unit test that checks the step where it creates n*instance total number of item requests.

Why does it create multiple requests in the same test?

> Recall that it does not know which type of request to make, so it creates all possible requests for all request types for that one item.

Within a test, where the policies and records are controlled, it should be possible to know what type of request is expected to be created.

I don't follow why that isn't possible here?

> Other tests in this file needed to make use of various fixtures so this test class has to extend APITests. We can create another class for true unit-level tests if needed.

I think splitting them out would be useful. I'm aware that due to the way the code is structured it is difficult to inject fakes to write more unit like tests. 

If we need a style of tests that doesn't use the API, but require infrastructure, then I think we should create that separately. It might be that is out of scope for this work (I'm saying this a lot at the moment, and the effect on the code base is noticeable).

Reply:
> Within a test, where the policies and records are controlled, it should be possible to know what type of request is expected to be created.

I don't follow why that isn't possible here?

This is not within the test we're talking about. This test checks the code that generates the fanout of requests. I could have used various validation code that mod-circulation already has to do preliminary validations so that we don't need to create a fanout list of requests to attempt to place, but I'd rather defer to the existing code to handle this logic than to try to piece them together.

Reply:
I think I still don’t follow. With a known set of items the output should be known?